File: hs/Day7.hs
#!runghc

module Main where

import Control.Monad
import Data.Maybe
import Text.ParserCombinators.Parsec hiding (count)

lexeme :: String -> Parser ()
lexeme s = spaces >> string s >> spaces

parseName :: Parser String
parseName = do
    name1 <- many1 letter
    space
    name2 <- many1 letter
    return $ name1 ++ " " ++ name2

parseContents :: Parser [(Int, String)]
parseContents = do
    try containsNothing <|> try containsThings
    where
        containsNothing = string "no other bags" >> return []
        containsThings = thing `sepBy1` (lexeme ",")
        thing = do
            n <- read <$> many digit
            spaces
            name <- parseName
            if n == 1 then
                lexeme "bag"
            else
                lexeme "bags"
            return (n, name)

parseLine :: Parser (String, [(Int, String)])
parseLine = do
    name <- parseName
    lexeme "bags contain"
    contents <- parseContents
    return (name, contents)

parse' :: String -> [Rule]
parse' str = 
    case parse parseLine "" str of
        Right (outer, inners) -> [Rule outer inner count | (count, inner) <- inners]
        e -> error $ show e

data Rule = Rule { outer :: String, inner :: String, count :: Int } deriving Show

inputs :: IO [Rule]
inputs = do
    lines' <- lines <$> readFile "inputs/Day7.txt"
    return $ join $ map parse' lines'

countUp :: String -> [Rule] -> [String]
countUp name rules =
    let parents = [outer rule | rule <- rules, inner rule == name] :: [String] in
        name : concatMap (flip countUp rules) parents

countDown :: String -> [Rule] -> Int
countDown name rules = 1 + sum [count rule * countDown (inner rule) rules | rule <- rules, outer rule == name]

for :: [a] -> (a -> b) -> [b]
for = flip map

uniq :: Eq a => [a] -> [a]
uniq [] = []
uniq (x:xs) | x `elem` xs = uniq xs
            | otherwise = x : uniq xs

main :: IO ()
main = do

    rules <- inputs

    print $ (length $ uniq $ countUp "shiny gold" rules) - 1
    print $ countDown "shiny gold" rules - 1

File: py/day7.py
#!python3

# from collections import namedtuple

# import pprint

RAW = open("../inputs/Day7.txt", "r").read()

rules = RAW.split("\n")


def rule_outer(rule):
    return rule.split(" ")[0] + " " + rule.split(" ")[1]


def rule_inner(rule):
    if "no other bags" in rule:
        return {}

    rule = rule.split("contain")[1]
    rule = rule.split(",")
    rule = [make_rule(r.strip()) for r in rule]

    return {
        k: v for k, v in rule
    }


def make_rule(rule):

    r = rule.split(" ")
    count = int(r[0])
    name = rule.split(" ")[1:3]
    name = " ".join(name)

    return (name, count)


def parse_rule(line):
    return {
        "outer_bag": rule_outer(line),
        "contents": rule_inner(line)
    }


def parse_rules(rules):

    for rule in rules:
        outer_color = rule_outer(rule)
        inner = rule_inner(rule)
        for (inner_color, count) in inner.items():
            yield (outer_color, inner_color, count)


def step1(rules):

    rules = list(parse_rules(rules))

    queue = ['shiny gold']
    visited = set()

    while len(queue) > 0:
        cur = queue.pop()
        visited.add(cur)
        for (outer, inner, _) in rules:
            if inner == cur:
                queue.append(outer)

    return len(visited) - 1


def step2(rules):

    def recur(parent):
        total = 1
        for (outer, inner, count) in rules:
            if outer == parent:
                total += count * recur(inner)

        return total

    rules = list(parse_rules(rules))

    return recur('shiny gold') - 1


print(step1(rules))
print(step2(rules))
